README.md
# sqlalchemy-aws-data-api
An Amazon Aurora Serverless Data API dialect for SQLAlchemy


Changes.rst
Changes for v0.5.0 (2023-12-29)
===============================

-  Update dependencies

-  Test and release infrastructure improvements

Changes for v0.4.1 (2022-05-18)
===============================

-  Support for fractional seconds when binding time and datetime data
   types (#40)

Changes for v0.4.0 (2022-02-27)
===============================

-  Bump dependency version

Changes for v0.3.4 (2022-01-23)
===============================

Bump dependency version

Changes for v0.3.3 (2022-01-23)
===============================

-  Update build tooling

Changes for v0.3.2 (2022-01-23)
===============================

-  Bump dependency to incorporate fix; add regression test

Changes for v0.3.1 (2021-12-25)
===============================

-  Update release script

Changes for v0.3.0 (2021-12-25)
===============================

-  Add supports_statement_cache flag. Fixes #29

-  Add error code extraction (#31)

-  Format with DataAPI supported format for dates (#24)

-  Set supports_sane_multi_rowcount = False on
   AuroraPostgresDataAPIDialect (#28)

-  Tell SQLAlchemy: mysql supports native decimal (fixes #25) (#26)

Changes for v0.2.7 (2020-12-12)
===============================

Bump dependency version

Changes for v0.2.6 (2020-11-13)
===============================

Bump dependency version

Changes for v0.2.5 (2020-10-03)
===============================

Bump dependency version

Changes for v0.2.4 (2020-10-02)
===============================

Bump dependency version

Changes for v0.2.3 (2020-10-02)
===============================

-  Bump dependency version

-  Fix strptime handling in Python 3.6 and earlier

Changes for v0.2.2 (2020-10-02)
===============================

-  Bump aurora-data-api dependency

Changes for v0.2.1 (2020-10-01)
===============================

-  Fall back to strptime if fromisoformat is not available

-  Merge pull request #8 from olinger/master (Add colspecs for mySQL
   date, time and datetime types)

-  Merge pull request #11 from romibuzi/master (Return error code
   instead of enum entry)

Changes for v0.2.0 (2020-01-02)
===============================

-  Bump aurora-data-api dependency

Changes for v0.1.6 (2020-01-02)
===============================

-  Add enum support

Changes for v0.1.5 (2020-01-01)
===============================

Fix handling of non-dialect-specific datetime types

Changes for v0.1.4 (2019-11-18)
===============================

-  Conform to dialect interface definition

-  MySQL: Return actual client charset

Changes for v0.1.3 (2019-11-10)
===============================

-  Begin MySQL support

Changes for v0.1.2 (2019-10-31)
===============================

-  Fix timestamp microsecond handling

Changes for v0.1.1 (2019-10-31)
===============================

-  Begin array support

-  Improve datetime support

Changes for v0.1.0 (2019-10-29)
===============================

-  Fix postgresql type compatibility issues

Changes for v0.0.2 (2019-10-24)
===============================

Add MySQL dialect

Changes for v0.0.1 (2019-10-10)
===============================

-  Begin sqlalchemy-aurora-data-api



test/test.py
import os
import sys
import unittest
import logging
import datetime
import enum
from uuid import UUID as uuid_type, uuid4

from sqlalchemy import (
    create_engine,
    Column,
    Integer,
    String,
    Boolean,
    Float,
    LargeBinary,
    Numeric,
    Date,
    Time,
    DateTime,
    Text,
    # Enum,
)
from sqlalchemy.dialects.postgresql import UUID, JSONB, JSON, DATE, TIME, TIMESTAMP, ARRAY
from sqlalchemy.orm import sessionmaker, declarative_base
from sqlalchemy.sql import text

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from sqlalchemy_aurora_data_api import register_dialects, _ADA_TIMESTAMP  # noqa

logging.basicConfig(level=logging.INFO)
logging.getLogger("aurora_data_api").setLevel(logging.DEBUG)
logging.getLogger("urllib3.connectionpool").setLevel(logging.DEBUG)

dialect_interface_attributes = {
    "name",
    "driver",
    "positional",
    "paramstyle",
    # "convert_unicode",
    # "encoding",
    "statement_compiler",
    "ddl_compiler",
    "server_version_info",
    "default_schema_name",
    "execution_ctx_cls",
    "execute_sequence_format",
    "preparer",
    "supports_alter",
    "max_identifier_length",
    # "supports_unicode_statements",
    # "supports_unicode_binds",
    "supports_sane_rowcount",
    "supports_sane_multi_rowcount",
    "preexecute_autoincrement_sequences",
    # "implicit_returning",
    "colspecs",
    "supports_default_values",
    "supports_sequences",
    "sequences_optional",
    "supports_native_enum",
    "supports_native_boolean",
    "dbapi_exception_translation_map",
}

dialect_interface_methods = {
    "connect",
    "create_connect_args",
    "create_xid",
    "denormalize_name",
    "do_begin",
    "do_begin_twophase",
    "do_close",
    "do_commit",
    "do_commit_twophase",
    "do_execute",
    "do_execute_no_params",
    "do_executemany",
    "do_prepare_twophase",
    "do_recover_twophase",
    "do_release_savepoint",
    "do_rollback",
    "do_rollback_to_savepoint",
    "do_rollback_twophase",
    "do_savepoint",
    "engine_created",
    "get_check_constraints",
    "get_columns",
    "get_dialect_cls",
    "get_foreign_keys",
    "get_indexes",
    "get_isolation_level",
    "get_pk_constraint",
    "get_table_comment",
    "get_table_names",
    "get_temp_table_names",
    "get_temp_view_names",
    "get_unique_constraints",
    "get_view_definition",
    "get_view_names",
    "has_sequence",
    "has_table",
    "initialize",
    "is_disconnect",
    "normalize_name",
    # "reflect_table",
    "reset_isolation_level",
    "set_isolation_level",
    "type_descriptor",
}

BasicBase = declarative_base()
Base = declarative_base()


class Socks(enum.Enum):
    red = 1
    green = 2
    black = 3


class BasicUser(BasicBase):
    __tablename__ = "sqlalchemy_aurora_data_api_testI"

    id = Column(Integer, primary_key=True)
    name = Column(String(64))
    fullname = Column(String(64))
    nickname = Column(String(64))
    birthday = Column(Date)
    eats_breakfast_at = Column(Time)
    married_at = Column(DateTime)


class User(Base):
    __tablename__ = "sqlalchemy_aurora_data_api_testJ"
    id = Column(Integer, primary_key=True)
    name = Column(String)
    fullname = Column(String)
    nickname = Column(String)
    doc = Column(JSONB)
    doc2 = Column(JSON)
    uuid = Column(UUID)
    uuid2 = Column(UUID(as_uuid=True), default=uuid4)
    flag = Column(Boolean, nullable=True)
    nonesuch = Column(Boolean, nullable=True)
    birthday = Column(DATE)
    wakes_up_at = Column(TIME)
    added = Column(TIMESTAMP)
    floated = Column(Float)
    nybbled = Column(LargeBinary)
    friends = Column(ARRAY(String))
    num_friends = Numeric(asdecimal=True)
    num_laptops = Numeric(asdecimal=False)
    first_date = Column(Date)
    note = Column(Text)
    # socks = Column(Enum(Socks))


class TestAuroraDataAPI(unittest.TestCase):
    @classmethod
    def tearDownClass(cls):
        pass

    def test_interface_conformance(self):
        for attr in dialect_interface_attributes:
            self.assertIn(attr, dir(self.engine.dialect))

        for attr in dialect_interface_methods:
            self.assertIn(attr, dir(self.engine.dialect))
            assert callable(getattr(self.engine.dialect, attr))


class TestAuroraDataAPIPostgresDialect(TestAuroraDataAPI):
    dialect = "postgresql+auroradataapi://"
    # dialect = "postgresql+psycopg2://" + getpass.getuser()

    @classmethod
    def setUpClass(cls):
        register_dialects()
        cls.db_name = os.environ.get("AURORA_DB_NAME", __name__)
        cls.engine = create_engine(cls.dialect + ":@/" + cls.db_name)

    def test_execute(self):
        with self.engine.connect() as conn:
            for result in conn.execute(text("select * from pg_catalog.pg_tables")):
                print(result)

    def test_orm(self):
        uuid = uuid4()
        doc = {"foo": [1, 2, 3]}
        blob = b"0123456789ABCDEF" * 1024
        friends = ["Scarlett O'Hara", 'Ada "Hacker" Lovelace']
        Base.metadata.create_all(self.engine)
        added = datetime.datetime.now().replace(microsecond=123456)
        ed_user = User(
            name="ed",
            fullname="Ed Jones",
            nickname="edsnickname",
            doc=doc,
            doc2=doc,
            uuid=str(uuid),
            flag=True,
            birthday=datetime.datetime.fromtimestamp(0),
            added=added,
            floated=1.2,
            nybbled=blob,
            friends=friends,
            num_friends=500,
            num_laptops=9000,
            first_date=added,
            note="note",
            # socks=Socks.red,
        )
        Session = sessionmaker(bind=self.engine)
        session = Session()

        session.query(User).delete()
        session.commit()

        session.add(ed_user)
        self.assertEqual(session.query(User).filter_by(name="ed").first().name, "ed")
        session.commit()
        self.assertGreater(session.query(User).filter(User.name.like("%ed")).count(), 0)
        u = session.query(User).filter(User.name.like("%ed")).first()
        self.assertEqual(u.doc, doc)
        self.assertEqual(u.doc2, doc)
        self.assertEqual(u.flag, True)
        self.assertEqual(u.nonesuch, None)
        self.assertEqual(u.birthday, datetime.date.fromtimestamp(0))
        self.assertEqual(u.added, added.replace(microsecond=123000))
        self.assertEqual(u.floated, 1.2)
        self.assertEqual(u.nybbled, blob)
        self.assertEqual(u.friends, friends)
        self.assertEqual(u.num_friends, 500)
        self.assertEqual(u.num_laptops, 9000)
        self.assertEqual(u.first_date, added.date())
        self.assertEqual(u.note, "note")
        print("FIXME: re-enable test for enums support", self.assertEqual, "u.socks", Socks.red)
        self.assertEqual(u.uuid, str(uuid))
        print("FIXME: re-enable test for uuid support", self.assertIsInstance, u.uuid2, uuid_type)

        # u.socks = Socks.green
        session.commit()

        session2 = Session()
        # u2 =
        session2.query(User).filter(User.name.like("%ed")).first()
        print("FIXME: re-enable test for enums support", self.assertEqual, "u2.socks", Socks.green)

    @unittest.skipIf(sys.version_info < (3, 7), "Skipping test that requires Python 3.7+")
    def test_timestamp_microsecond_padding(self):
        ts = "2019-10-31 09:37:17.3186"
        processor = _ADA_TIMESTAMP.result_processor(_ADA_TIMESTAMP, None, None)
        self.assertEqual(processor(ts), datetime.datetime.fromisoformat(ts.ljust(26, "0")))


class TestAuroraDataAPIMySQLDialect(TestAuroraDataAPI):
    dialect = "mysql+auroradataapi://"

    @classmethod
    def setUpClass(cls):
        register_dialects()
        cls.db_name = os.environ.get("AURORA_DB_NAME", __name__)
        cls.engine = create_engine(cls.dialect + ":@/" + cls.db_name + "?charset=utf8mb4")

    def test_execute(self):
        with self.engine.connect() as conn:
            for result in conn.execute("select * from information_schema.tables"):
                print(result)

    def test_orm(self):
        BasicBase.metadata.create_all(self.engine)
        birthday = datetime.datetime.fromtimestamp(0).date()
        eats_breakfast_at = datetime.time(9, 0, 0, 123)
        married_at = datetime.datetime(2020, 2, 20, 2, 20, 2, 200200)
        ed_user = BasicUser(
            name="ed",
            fullname="Ed Jones",
            nickname="edsnickname",
            birthday=birthday,
            eats_breakfast_at=eats_breakfast_at,
            married_at=married_at,
        )
        Session = sessionmaker(bind=self.engine)
        session = Session()

        session.query(BasicUser).delete()
        session.commit()

        session.add(ed_user)
        self.assertEqual(session.query(BasicUser).filter_by(name="ed").first().name, "ed")
        session.commit()
        self.assertGreater(session.query(BasicUser).filter(BasicUser.name.like("%ed")).count(), 0)
        u = session.query(BasicUser).filter(BasicUser.name.like("%ed")).first()
        self.assertEqual(u.nickname, "edsnickname")
        self.assertEqual(u.birthday, birthday)
        self.assertEqual(u.eats_breakfast_at, eats_breakfast_at.replace(microsecond=0))
        self.assertEqual(u.married_at, married_at.replace(microsecond=0))


if __name__ == "__main__":
    unittest.main()


searchindex.js
Search.setIndex({"docnames": ["index"], "filenames": ["index.rst"], "titles": ["sqlalchemy-aurora-data-api - An AWS Aurora Serverless Data API dialect for SQLAlchemy"], "terms": {"thi": 0, "packag": 0, "provid": 0, "access": 0, "postgresql": 0, "mysql": 0, "databas": 0, "via": 0, "pip": 0, "set": 0, "up": 0, "cluster": 0, "enabl": 0, "If": 0, "you": 0, "have": 0, "previous": 0, "can": 0, "follow": 0, "cli": 0, "command": 0, "rd": 0, "modifi": 0, "db": 0, "identifi": 0, "db_cluster_nam": 0, "http": 0, "endpoint": 0, "appli": 0, "immedi": 0, "save": 0, "credenti": 0, "secret": 0, "manag": 0, "us": 0, "format": 0, "expect": 0, "json": 0, "object": 0, "kei": 0, "usernam": 0, "password": 0, "secretsmanag": 0, "creat": 0, "name": 0, "my_db": 0, "put": 0, "valu": 0, "id": 0, "string": 0, "jq": 0, "n": 0, "env": 0, "pguser": 0, "pgpassword": 0, "configur": 0, "your": 0, "line": 0, "standard": 0, "convent": 0, "verifi": 0, "everyth": 0, "work": 0, "correctli": 0, "run": 0, "test": 0, "queri": 0, "execut": 0, "statement": 0, "resourc": 0, "arn": 0, "resource_arn": 0, "secret_arn": 0, "sql": 0, "select": 0, "from": 0, "pg_catalog": 0, "pg_tabl": 0, "here": 0, "refer": 0, "which": 0, "found": 0, "consol": 0, "click": 0, "describ": 0, "abov": 0, "when": 0, "deploi": 0, "code": 0, "ec2": 0, "instanc": 0, "ec": 0, "ek": 0, "contain": 0, "lambda": 0, "iam": 0, "polici": 0, "amazonrdsdatafullaccess": 0, "grant": 0, "role": 0, "permiss": 0, "while": 0, "i": 0, "conveni": 0, "we": 0, "recommend": 0, "own": 0, "scope": 0, "down": 0, "least": 0, "privileg": 0, "product": 0, "applic": 0, "The": 0, "regist": 0, "two": 0, "auroradataapi": 0, "create_engin": 0, "connect_arg": 0, "keyword": 0, "argument": 0, "ar": 0, "requir": 0, "connect": 0, "aurora_cluster_arn": 0, "also": 0, "resourcearn": 0, "given": 0, "specifi": 0, "environ": 0, "variabl": 0, "aurora_secret_arn": 0, "all": 0, "content": 0, "other": 0, "than": 0, "protocol": 0, "path": 0, "compon": 0, "my_db_nam": 0, "exampl": 0, "below": 0, "ignor": 0, "import": 0, "cluster_arn": 0, "u": 0, "east": 0, "1": 0, "123456789012": 0, "my": 0, "engin": 0, "echo": 0, "true": 0, "dict": 0, "conn": 0, "result": 0, "print": 0, "between": 0, "sophist": 0, "featur": 0, "tunnel": 0, "over": 0, "ha": 0, "advantag": 0, "context": 0, "It": 0, "elimin": 0, "need": 0, "open": 0, "port": 0, "public": 0, "ip": 0, "address": 0, "pool": 0, "stateless": 0, "instead": 0, "state": 0, "intern": 0, "tcp": 0, "most": 0, "driver": 0, "becom": 0, "invalid": 0, "after": 0, "go": 0, "through": 0, "freez": 0, "thaw": 0, "cycl": 0, "caus": 0, "error": 0, "burden": 0, "server": 0, "abandon": 0, "base": 0, "authent": 0, "handl": 0, "directli": 0, "python": 0, "log": 0, "To": 0, "output": 0, "level": 0, "basicconfig": 0, "getlogg": 0, "aurora_data_api": 0, "setlevel": 0, "project": 0, "home": 0, "page": 0, "github": 0, "read": 0, "doc": 0, "distribut": 0, "pypi": 0, "chang": 0, "2": 0, "0": 0, "client": 0, "depend": 0, "pleas": 0, "report": 0, "issu": 0, "request": 0, "etc": 0, "under": 0, "term": 0, "apach": 0, "version": 0}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"sqlalchemi": 0, "aurora": 0, "data": 0, "api": 0, "an": 0, "aw": 0, "serverless": 0, "dialect": 0, "instal": 0, "prerequisit": 0, "usag": 0, "motiv": 0, "debug": 0, "link": 0, "bug": 0, "licens": 0, "document": 0}, "envversion": {"sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 60}, "alltitles": {"sqlalchemy-aurora-data-api - An AWS Aurora Serverless Data API dialect for SQLAlchemy": [[0, "sqlalchemy-aurora-data-api-an-aws-aurora-serverless-data-api-dialect-for-sqlalchemy"]], "Installation": [[0, "installation"]], "Prerequisites": [[0, "prerequisites"]], "Usage": [[0, "usage"]], "Motivation": [[0, "motivation"]], "Debugging": [[0, "debugging"]], "Links": [[0, "links"]], "Bugs": [[0, "bugs"]], "License": [[0, "license"]], "API documentation": [[0, "api-documentation"]]}, "indexentries": {}})

sqlalchemy_aurora_data_api/__init__.py
"""
sqlalchemy-aurora-data-api
"""

import json, datetime, re

from sqlalchemy import cast, func, util
import sqlalchemy.sql.sqltypes as sqltypes
from sqlalchemy.dialects.postgresql.base import PGDialect
from sqlalchemy.dialects.postgresql import JSON, JSONB, UUID, DATE, TIME, TIMESTAMP, ARRAY, ENUM
from sqlalchemy.dialects.mysql.base import MySQLDialect

import aurora_data_api


class _ADA_SA_JSON(sqltypes.JSON):
    def bind_expression(self, value):
        return cast(value, sqltypes.JSON)


class _ADA_JSON(JSON):
    def bind_expression(self, value):
        return cast(value, JSON)


class _ADA_JSONB(JSONB):
    def bind_expression(self, value):
        return cast(value, JSONB)


class _ADA_UUID(UUID):
    def bind_expression(self, value):
        return cast(value, UUID)


class _ADA_ENUM(ENUM):
    def bind_expression(self, value):
        return cast(value, self)


# TODO: is TZ awareness needed here?
class _ADA_DATETIME_MIXIN:
    iso_ts_re = re.compile(r"\d{4}-\d\d-\d\d \d\d:\d\d:\d\d\.\d+")

    @staticmethod
    def ms(value):
        # Three digit fractional second component, truncated and zero padded. This is what the data api requires.
        return str(value.microsecond).zfill(6)[:-3]

    def bind_processor(self, dialect):
        def process(value):
            return value.isoformat() if isinstance(value, self.py_type) else value

        return process

    def bind_expression(self, value):
        return cast(value, self.sa_type)

    def result_processor(self, dialect, coltype):
        def process(value):
            # When the microsecond component ends in zeros, they are omitted from the return value,
            # and datetime.datetime.fromisoformat can't parse the result (example: '2019-10-31 09:37:17.31869
            # '). Pad it.
            if isinstance(value, str) and self.iso_ts_re.match(value):
                value = self.iso_ts_re.sub(lambda match: match.group(0).ljust(26, "0"), value)
            if isinstance(value, str):
                try:
                    return self.py_type.fromisoformat(value)
                except AttributeError:  # fromisoformat not supported on Python < 3.7
                    if self.py_type == datetime.date:
                        return datetime.datetime.strptime(value, "%Y-%m-%d").date()
                    if self.py_type == datetime.time:
                        return datetime.datetime.strptime(value, "%H:%M:%S").time()
                    if "." in value:
                        return datetime.datetime.strptime(value, "%Y-%m-%d %H:%M:%S.%f")
                    return datetime.datetime.strptime(value, "%Y-%m-%d %H:%M:%S")
            return value

        return process


class _ADA_DATE(_ADA_DATETIME_MIXIN, DATE):
    py_type = datetime.date
    sa_type = sqltypes.Date

    def bind_processor(self, dialect):
        def process(value):
            return value.strftime("%Y-%m-%d") if isinstance(value, self.py_type) else value

        return process


class _ADA_TIME(_ADA_DATETIME_MIXIN, TIME):
    py_type = datetime.time
    sa_type = sqltypes.Time

    def bind_processor(self, dialect):
        def process(value):
            return value.strftime("%H:%M:%S.") + self.ms(value) if isinstance(value, self.py_type) else value

        return process


class _ADA_TIMESTAMP(_ADA_DATETIME_MIXIN, TIMESTAMP):
    py_type = datetime.datetime
    sa_type = sqltypes.DateTime

    def bind_processor(self, dialect):
        def process(value):
            return value.strftime("%Y-%m-%d %H:%M:%S.") + self.ms(value) if isinstance(value, self.py_type) else value

        return process


class _ADA_ARRAY(ARRAY):
    def bind_processor(self, dialect):
        def process(value):
            # FIXME: escape strings properly here
            return "\v".join(value) if isinstance(value, list) else value

        return process

    def bind_expression(self, value):
        return func.string_to_array(value, "\v")


class AuroraMySQLDataAPIDialect(MySQLDialect):
    # See https://docs.sqlalchemy.org/en/13/core/internals.html#sqlalchemy.engine.interfaces.Dialect
    driver = "aurora_data_api"
    default_schema_name = None
    supports_native_decimal = True
    colspecs = util.update_copy(
        MySQLDialect.colspecs,
        {
            sqltypes.Date: _ADA_DATE,
            sqltypes.Time: _ADA_TIME,
            sqltypes.DateTime: _ADA_TIMESTAMP,
        },
    )
    supports_statement_cache = True

    @classmethod
    def import_dbapi(cls):
        return aurora_data_api

    def _detect_charset(self, connection):
        return connection.execute("SHOW VARIABLES LIKE 'character_set_client'").fetchone()[1]

    def _extract_error_code(self, exception):
        return exception.args[0].value


class AuroraPostgresDataAPIDialect(PGDialect):
    # See https://docs.sqlalchemy.org/en/13/core/internals.html#sqlalchemy.engine.interfaces.Dialect
    driver = "aurora_data_api"
    default_schema_name = None
    colspecs = util.update_copy(
        PGDialect.colspecs,
        {
            sqltypes.JSON: _ADA_SA_JSON,
            JSON: _ADA_JSON,
            JSONB: _ADA_JSONB,
            UUID: _ADA_UUID,
            sqltypes.Date: _ADA_DATE,
            sqltypes.Time: _ADA_TIME,
            sqltypes.DateTime: _ADA_TIMESTAMP,
            sqltypes.Enum: _ADA_ENUM,
            ARRAY: _ADA_ARRAY,
        },
    )
    supports_sane_multi_rowcount = False
    supports_statement_cache = True

    @classmethod
    def import_dbapi(cls):
        return aurora_data_api

    def _extract_error_code(self, exception):
        return exception.args[0].value


def register_dialects():
    from sqlalchemy.dialects import registry

    registry.register("mysql.auroradataapi", __name__, AuroraMySQLDataAPIDialect.__name__)
    registry.register("postgresql.auroradataapi", __name__, AuroraPostgresDataAPIDialect.__name__)
